From 2ac37f7a5387571f8ed14d6689d62a73687cf1ab Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Pierret=20=28fepitre=29?=
 <frederic.epitre@orange.fr>
Date: Sat, 10 Nov 2018 19:10:07 +0100
Subject: [PATCH] check for Qubes OS hardware required features

Warn if the hardware lacks features required for proper Qubes OS operation.

Rework based on previous commits 696bd4c, 63043751a, 7489992dd, 3791fd7 and f6bfe11ab.
---
 pyanaconda/core/constants.py |  6 ++++--
 pyanaconda/core/util.py      | 29 +++++++++++++++++++++++++----
 2 files changed, 29 insertions(+), 6 deletions(-)

diff --git a/pyanaconda/core/constants.py b/pyanaconda/core/constants.py
index ece934648..161e5575a 100644
--- a/pyanaconda/core/constants.py
+++ b/pyanaconda/core/constants.py
@@ -146,8 +146,10 @@ WARNING_SUPPORT_REMOVED = N_(
 )
 
 WARNING_HARDWARE_UNSUPPORTED = N_(
-    "This hardware (or a combination thereof) is not supported by Red Hat. For more information "
-    "on supported hardware, please refer to http://www.redhat.com/hardware."
+    "This hardware lacks features required by Qubes OS. Missing features: %(features)s. Without "
+    "these features, Qubes OS will not function normally. It is recommended that only developers "
+    "and power users proceed with the installation. For more information on supported "
+    "hardware, please refer to https://www.qubes-os.org/doc/system-requirements/"
 )
 
 # Password type
diff --git a/pyanaconda/core/util.py b/pyanaconda/core/util.py
index 6106671c8..7043bcc24 100644
--- a/pyanaconda/core/util.py
+++ b/pyanaconda/core/util.py
@@ -989,10 +989,31 @@ def detect_unsupported_hardware(install_class):
         return []
 
     # Check TAINT_HARDWARE_UNSUPPORTED
-    if not install_class.detect_unsupported_hardware:
-        log.debug("%s doesn't support TAINT_HARDWARE_UNSUPPORTED.", install_class.name)
-    elif get_kernel_taint(TAINT_HARDWARE_UNSUPPORTED):
-        warnings.append(WARNING_HARDWARE_UNSUPPORTED)
+    try:
+        xl_info = subprocess.check_output(['xl', 'info'])
+        xl_dmesg = subprocess.check_output(['xl', 'dmesg'])
+    except subprocess.CalledProcessError:
+        warnings.append('Unable to check hardware support: xl call failed')
+    else:
+        missing_features = []
+        for line in xl_info.splitlines():
+            if line.startswith(b'virt_caps'):
+                if b'hvm' not in line:
+                    missing_features.append('HVM/VT-x/AMD-V')
+                if b'hvm_directio' not in line:
+                    missing_features.append('IOMMU/VT-d/AMD-Vi')
+
+        if b'HVM: Hardware Assisted Paging (HAP) detected' not in xl_dmesg:
+            missing_features.append('HAP/SLAT/EPT/RVI')
+
+        # slightly different wording for Intel and AMD
+        if b'Intel VT-d Interrupt Remapping enabled' not in xl_dmesg \
+                and b'Interrupt remapping enabled' not in xl_dmesg:
+            missing_features.append('Interrupt Remapping')
+
+        if missing_features:
+            status = ', '.join(missing_features)
+            warnings.append(WARNING_HARDWARE_UNSUPPORTED % {'features': status})
 
     # Check TAINT_SUPPORT_REMOVED
     if not install_class.detect_support_removed:
-- 
2.20.1

